// react-context-router/dist/Link.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var Link = function Link(_ref, _ref2) {
  var href = _ref.href;

  var props = _objectWithoutProperties(_ref, ['href']);

  var history = _ref2.history;
  var location = _ref2.location;

  var handleClick = function handleClick(e) {
    e.preventDefault();
    history.push(href);
  };

  return _react2.default.createElement('a', _extends({}, props, {
    href: href,
    onClick: handleClick }));
};

Link.contextTypes = {
  history: _react2.default.PropTypes.object,
  location: _react2.default.PropTypes.object
};

exports.default = Link;
